033.Search for a Range/033.search for a range.md

```diff
@@ -0,0 +1,87 @@
+# 035.Search for a Range 
+
+**<font color=red>level: Easy</font>**
+
+## Description
+
+Given a sorted array of integers, find the starting and ending position of a given target value.
+Your algorithm’s runtime complexity must be in the order of O(log n).
+If the target is not found in the array, return [-1, -1].
+For example, Given [5, 7, 7, 8, 8, 10] and target value 8, return [3, 4].
+
+## Solution in C++
+
+* Approach 1
+
+```c++
+class Solution {
+public:
+	vector<int> searchRange(int A[], int n, int target) {
+		const int l = distance(A, lower_bound(A, A + n, target));
+		const int u = distance(A, prev(upper_bound(A, A + n, target)));
+		if (A[l] != target) // not found
+			return vector<int> { -1, -1 };
+		else
+			return vector<int> { l, u };
+		}
+};
+
+```
+* Approach 2 
+
+```c++
+
+int lower_bound(vector<int>& arr, int key) {
+    int half;
+    int len = arr.size();
+    int mid;
+    int first = 0;
+    while (len > 0) {
+        half = len >> 1;
+        mid = first + half;
+        //in the right part
+        if (arr[mid] < key) {
+            first = mid + 1;
+            //because first=mid+1,so len = len - half - 1
+            len = len - half - 1;
+        } else {
+            //in the left part
+            len = half;
+        }
+    }
+    return first;
+}
+int upper_bound(vector<int>& arr, int key) {
```

Comment:
I think lowerbound and upperbound have the same logic. 
We should reduce duplicate codes, right?

Reply:
Maybe we should rewrite it like this:
```c++
int lower_bound(vector<int>& arr, int key, const std::function<> &comparer) {
//xxxxxx
if (comparer(arr[mid], key)) {
//xxxxxx
}
/xxxxxxx
}
```